src/main/java/bio/terra/service/snapshot/flight/create/CreateSnapshotPrimaryDataQueryInterface.java

```diff
@@ -1,39 +1,77 @@
 package bio.terra.service.snapshot.flight.create;
 
+import bio.terra.common.iam.AuthenticatedUserRequest;
 import bio.terra.grammar.Query;
 import bio.terra.grammar.exception.InvalidQueryException;
+import bio.terra.model.SnapshotAccessRequestResponse;
+import bio.terra.model.SnapshotRequestContentsModel;
 import bio.terra.model.SnapshotRequestModel;
 import bio.terra.model.SnapshotRequestQueryModel;
 import bio.terra.service.common.CommonFlightUtils;
 import bio.terra.service.dataset.AssetSpecification;
 import bio.terra.service.dataset.Dataset;
 import bio.terra.service.dataset.DatasetService;
 import bio.terra.service.snapshot.Snapshot;
+import bio.terra.service.snapshot.SnapshotDao;
 import bio.terra.service.snapshot.exception.AssetNotFoundException;
+import bio.terra.service.snapshotbuilder.SnapshotBuilderService;
+import bio.terra.service.snapshotbuilder.SnapshotRequestDao;
 import bio.terra.stairway.FlightContext;
 import bio.terra.stairway.StepResult;
 import java.time.Instant;
 import java.util.List;
+import java.util.UUID;
 
 public interface CreateSnapshotPrimaryDataQueryInterface {
 
   default StepResult prepareQueryAndCreateSnapshot(
       FlightContext context,
       Snapshot snapshot,
       SnapshotRequestModel snapshotReq,
-      DatasetService datasetService)
+      DatasetService datasetService,
+      SnapshotBuilderService snapshotBuilderService,
+      SnapshotRequestDao snapshotRequestDao,
+      SnapshotDao snapshotDao,
+      AuthenticatedUserRequest userReq)
       throws InterruptedException {
-    SnapshotRequestQueryModel snapshotQuerySpec = snapshotReq.getContents().get(0).getQuerySpec();
+    AssetSpecification assetSpecification;
+    String sqlQuery;
+    Instant createdAt;
 
-    Query query = Query.parse(snapshotQuerySpec.getQuery());
-    String datasetName = query.getDatasetName();
-    Dataset dataset = datasetService.retrieveByName(datasetName);
-    AssetSpecification assetSpecification =
-        retrieveAssetSpecification(dataset, snapshotQuerySpec.getAssetName());
-    validateRootTable(query, assetSpecification);
-    String sqlQuery = translateQuery(query, dataset);
+    if (snapshotReq
+        .getContents()
+        .get(0)
+        .getMode()
+        .equals(SnapshotRequestContentsModel.ModeEnum.BYREQUESTID)) {
+      UUID accessRequestId =
+          snapshotReq.getContents().get(0).getRequestIdSpec().getSnapshotRequestId();
+      SnapshotAccessRequestResponse accessRequest = snapshotRequestDao.getById(accessRequestId);
+
+      UUID dataReleaseSnapshotId = accessRequest.getSourceSnapshotId();
+      Snapshot dataReleaseSnapshot = snapshotDao.retrieveSnapshot(dataReleaseSnapshotId);
+      // get the underlying dataset for the snapshot
+      if (dataReleaseSnapshot.getSnapshotSources().isEmpty()) {
+        throw new IllegalArgumentException("Snapshot does not have a source dataset");
+      }
+      Dataset dataset = dataReleaseSnapshot.getSnapshotSources().get(0).getDataset();
```

Comment:
This would avoid the if check above:
```suggestion
      Dataset dataset = dataReleaseSnapshot.getSnapshotSources().stream().findFirst().orElseThrow(() -> new IllegalArgumentException("Snapshot does not have a source dataset"));
```